File: Assets/Resources/Skill/BlizzardCaster.js
#pragma strict
var skillIndex: int = 1;
var skillName: String = 'Blizzard';
var coolDownTime: double = 15000;
var guidingTime: double = 10000;
var isGuiding: boolean = false;
var owner: GameObject;
var enemy: GameObject;
private var epochStart: System.DateTime = new System.DateTime(1970, 1, 1, 0, 0, 0, System.DateTimeKind.Utc);
private var eventCenter: EventCenter;
private var isButtonPressed: boolean = false;
private var timeStartCooling: double = 0;
private var timeStartGuiding: double = 0;
private var blizzard: GameObject;

function Awake () {
	eventCenter = GameObject.FindWithTag('EventCenter').GetComponent(EventCenter);
	eventCenter.RegisterListener(eventCenter, 'skillButtonDown', gameObject, OnSkillButtonDown);
	eventCenter.RegisterListener(eventCenter, 'skillButtonUp', gameObject, OnSkillButtonUp);
	eventCenter.RegisterListener(eventCenter, 'leftButtonPressed', gameObject, OnPlayerMove);
	eventCenter.RegisterListener(eventCenter, 'rightButtonPressed', gameObject, OnPlayerMove);
}
function Update () {
	if(false == isButtonPressed) {	
		if(true == isGuiding) {
			StopGuiding();
			isGuiding = false;
		}
		return;
	}
	var timestamp: double = (System.DateTime.UtcNow - epochStart).TotalMilliseconds;
	if(false == isGuiding) {
		if(true == GetIsCoolDownFinished()) {
			isGuiding = true;
			timeStartGuiding = timestamp;
			StartGuiding();
			StartCoolDown();
		}
	} else {
		if((timeStartGuiding + guidingTime) < timestamp) {
			isGuiding = false;
			StopGuiding();
		}
	}
}
var OnSkillButtonDown = function(e: SbiEvent) {
	var index: int = System.Convert.ToInt32(e.data);
	if(skillIndex != index) return;
	isButtonPressed = true;
};
var OnSkillButtonUp = function(e: SbiEvent) {
	var index: int = System.Convert.ToInt32(e.data);
	if(skillIndex != index) return;
	isButtonPressed = false;
};
var OnPlayerMove = function(e: SbiEvent) {
	isGuiding = false;
	isButtonPressed = false;
};
private function StartGuiding() {
	var targetPosition: Vector3 = transform.position;
	if(null != enemy) { targetPosition = enemy.transform.position; }
	if(null == blizzard) {
		blizzard = Instantiate(
			Resources.Load('Skill/Blizzard'), targetPosition, transform.rotation
		) as GameObject;
	} else {
		blizzard.transform.position = targetPosition;
	}
	blizzard.SetActive(true);
}
private function StopGuiding() { blizzard.SetActive(false); }
private function StartCoolDown() {
	timeStartCooling = (System.DateTime.UtcNow - epochStart).TotalMilliseconds;
}
private function GetIsCoolDownFinished(): boolean {
	var timestamp: double = (System.DateTime.UtcNow - epochStart).TotalMilliseconds;
	return (timeStartCooling + coolDownTime) < timestamp;
}

File: Assets/Script/Skill/BlizzardCaster.cs
using UnityEngine;

public class BlizzardCaster : MonoBehaviour {
	public int skillIndex = 1;
	public string skillName = "Blizzard";
	public GameObject owner;
	public GameObject Owner {
		get {
			if(!owner && transform.parent && transform.parent.parent) {
				owner = transform.parent.parent.gameObject;
			}
			return owner;
		}
	}

	private SkillProperties skillProperties;
	private bool isButtonPressed = false;
	private CoolDownTimer coolDownTimer;
	private GuideTimer guideTimer;
	private Blizzard blizzard;

	public void Awake () {
		skillProperties = GetComponent<SkillProperties>();
		coolDownTimer = GetComponent<CoolDownTimer>();
		guideTimer = GetComponent<GuideTimer>();
		EventManager eventManager = EventManager.Instance;
		eventManager.RegisterListener(eventManager, "skillButtonDown", gameObject, OnSkillButtonDown);
		eventManager.RegisterListener(eventManager, "skillButtonUp", gameObject, OnSkillButtonUp);
		eventManager.RegisterListener(eventManager, "leftButtonPressed", gameObject, OnPlayerMove);
		eventManager.RegisterListener(eventManager, "rightButtonPressed", gameObject, OnPlayerMove);
		eventManager.RegisterListener(eventManager, "startGuiding", gameObject, OnStartGuiding);
		eventManager.RegisterListener(eventManager, "stopGuiding", gameObject, OnStopGuiding);
	}
	public void Update () {
		if(GameManager.Instance.GetPlayerCharacter() != owner) { return; }
		if(false == isButtonPressed) {	
			if(true == guideTimer.isGuiding) {
				StopGuidingRPC();
			}
			return;
		}
		if(false == guideTimer.isGuiding) {
			if(true == coolDownTimer.GetIsCoolDownFinished()) {
				coolDownTimer.StartCoolDown();
				StartGuidingRPC();
			}
		} else {
			if(true == guideTimer.GetIsGuidingFinished()) {
				StopGuidingRPC();
			}
		}
	}
	public void OnSkillButtonDown(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		SkillButtonEventData data = e.data as SkillButtonEventData;
		if(skillIndex != data.index) {
			if(true == guideTimer.isGuiding) { StopGuidingRPC(); }
			return;
		}
		isButtonPressed = true;
	}
	public void OnSkillButtonUp(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		SkillButtonEventData data = e.data as SkillButtonEventData;
		if(skillIndex != data.index) return;
		isButtonPressed = false;
	}
	public void OnPlayerMove(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		isButtonPressed = false;
		StopGuidingRPC();
	}
	private void StartGuidingRPC() {
		if(null == gameObject.transform.parent) { return; }
		SkillHandler handler = gameObject.transform.parent.gameObject.GetComponent<SkillHandler>();
		handler.StartGuidingRPC(skillProperties.skillId);
	}
	private void StopGuidingRPC() {
		if(null == gameObject.transform.parent) { return; }
		guideTimer.StopGuiding();
		SkillHandler handler = gameObject.transform.parent.gameObject.GetComponent<SkillHandler>();
		handler.StopGuidingRPC(skillProperties.skillId);
	}
	private void OnStartGuiding(SbiEvent e) {
		GuidingEventData data = (GuidingEventData)e.data;
		if((Owner != data.role) || (skillProperties.skillId != data.skillId)) {
			return;
		}
		guideTimer.StartGuiding();

        GameObject target = Owner.GetComponent<LookAt>().target;
		Vector3 targetPosition = transform.position;
		if(null != target) { targetPosition = target.transform.position; }

		if(null == blizzard) {
			GameObject blizzardGameObject = Instantiate(
				Resources.Load("Prefab/Skill/Blizzard"), targetPosition, transform.rotation
			) as GameObject;
			blizzardGameObject.GetComponent<Faction>().SetFaction(Owner.GetComponent<Faction>());
			blizzard = blizzardGameObject.GetComponent<Blizzard>();
			blizzard.owner = Owner;
		} else {
			blizzard.transform.position = targetPosition;
		}
		blizzard.gameObject.SetActive(true);
	}
	private void OnStopGuiding(SbiEvent e) {
		GuidingEventData data = (GuidingEventData)e.data;
		if((Owner != data.role) || (skillProperties.skillId != data.skillId)) {
			return;
		}
		guideTimer.StopGuiding();
		blizzard.gameObject.SetActive(false);
	}
}
